src/test/spec/unified_runner/mod.rs

```diff
@@ -0,0 +1,328 @@
+mod test_file;
+
+use std::collections::HashMap;
+
+use futures::stream::TryStreamExt;
+use lazy_static::lazy_static;
+use semver::Version;
+
+use crate::{
+    bson::{doc, Document},
+    concern::{Acknowledgment, WriteConcern},
+    options::{
+        CollectionOptions,
+        FindOptions,
+        InsertManyOptions,
+        ReadConcern,
+        ReadPreference,
+        SelectionCriteria,
+    },
+    test::{assert_matches, run_spec_test, util::EventClient, TestClient, LOCK},
+};
+
+pub use self::{
+    entity::Entity,
+    matcher::results_match,
+    operation::EntityOperation,
+    test_event::TestEvent,
+    test_file::{CollectionData, Entity as TestFileEntity, ExpectError, Operation, TestFile},
+};
+
+lazy_static! {
+    static ref SPEC_VERSIONS: Vec<Version> = vec![Version::parse("1.0.0").unwrap()];
+}
+
+pub async fn run_unified_format_test(test_file: TestFile) {
+    let version_matches = SPEC_VERSIONS.iter().any(|req| {
+        if req.major != test_file.schema_version.major {
+            return false;
+        }
+        if req.minor < test_file.schema_version.minor {
+            return false;
+        }
+        // patch versions do not affect the test file format
+        true
+    });
+    if !version_matches {
+        println!(
+            "Test runner not compatible with specification version {}",
+            &test_file.schema_version
+        );
+        return;
+    }
+
+    let client = TestClient::new().await;
+    if let Some(requirements) = test_file.run_on_requirements {
+        if !requirements
+            .iter()
+            .any(|requirement| requirement.can_run_on(&client))
+        {
+            println!("Client topology not compatible with test");
+            return;
+        }
+    }
+
+    for test_case in test_file.tests {
+        if let Some(skip_reason) = test_case.skip_reason {
+            println!("skipping {}: {}", &test_case.description, skip_reason);
+            return;
+        }
+
+        if let Some(requirements) = test_case.run_on_requirements {
+            if !requirements
+                .iter()
+                .any(|requirement| requirement.can_run_on(&client))
+            {
+                println!(
+                    "{}: client topology not compatible with test",
+                    &test_case.description
+                );
+                return;
+            }
+        }
+
+        if let Some(ref initial_data) = test_file.initial_data {
+            for data in initial_data {
+                insert_initial_data(data, &client).await;
+            }
+        }
+
+        let mut entities: HashMap<String, Entity> = HashMap::new();
+        if let Some(ref create_entities) = test_file.create_entities {
+            populate_entity_map(&mut entities, create_entities).await;
+        }
+
+        let mut failpoint_disable_commands: Vec<Document> = Vec::new();
+        for operation in test_case.operations {
+            if operation.object.as_str() == "testRunner" {
```

Comment:
Updated to add an `execute_on_test_runner` method.

Reply:
We're still switching on the type of object when executing the operation, which I don't believe is technically correct. There doesn't seem to be anything in the spec that precludes the "Special Test Operations" (i.e. the test runner ones) from making use of the `exectResult`, `expectError`, or `saveResultAsEntity` fields (or any shared fields in the future, for that matter), so ideally we'd aim to take the same code path for executing both types of operations, regardless of their target, for the sake of forwards compatibility. Right now only the entity operations have logic for using those fields.

To that end, I think instead of separately calling `entity.execute_operation` and `operation.execute_on_test_runner`, we could have a single `Operation::execute(&self, runner: &mut TestRunner) -> Result<Option<Bson>>` that is called regardless of operation type. Then we can remove the early branching on `OperationObject` in the runner and leave that internal to the implementation of `Operation::execute`.

I imagine `TestRunner` could look something like:
```rust
struct TestRunner {
    internal_client: TestClient,
    entities: HashMap<String, Entity>,
    failpoint_disable_commands: Vec<Document>,
}
```
This would also allow operations like `TargetedFailPoint` to update `failpoint_disable_commands` directly, rather than returning it from `execute_on_test_runner`, which I think would be too brittle in the face of future changes.

What do you think?